File: ffi/src/logger.cpp
#include "logger.hpp"

class Logger : public libfreenect2::Logger {
 public:
  explicit Logger(rust::Fn<void(LogLevel, const rust::String &)> log_fn)
      : log_fn(log_fn) {}

  Level level() const override {
    return Level::Debug;
  }

  void log(Level level, const std::string &message) override {
    log_fn(static_cast<LogLevel>(level), message);
  }

 private:
  rust::Fn<void(LogLevel, const rust::String &)> log_fn;
};

LIBFREENECT2_MAYBE_UNUSED void libfreenect2_ffi::create_logger(
    rust::Fn<void(LogLevel, const rust::String &)> log_fn) {
  auto logger = new Logger(log_fn);
  libfreenect2::setGlobalLogger(logger);
}

File: libfreenect2-rs/src/util/logger.rs
use crate::ffi::libfreenect2::create_logger;
use std::sync::Once;

static LOGGER_INIT: Once = Once::new();

pub(crate) fn init_logger() {
  LOGGER_INIT.call_once(|| {
    create_logger(|level, message| match level {
      crate::ffi::libfreenect2::LogLevel::Error => log::error!("{}", message),
      crate::ffi::libfreenect2::LogLevel::Warning => log::warn!("{}", message),
      crate::ffi::libfreenect2::LogLevel::Info => log::info!("{}", message),
      crate::ffi::libfreenect2::LogLevel::Debug => log::debug!("{}", message),
      _ => log::error!("Unknown log level: {}", message),
    })
    .unwrap();
  });
}
